controllers/metal3.io/host_state_machine.go

```diff
@@ -226,6 +230,34 @@ func (hsm *hostStateMachine) checkInitiateDelete() bool {
 	return true
 }
 
+func (hsm *hostStateMachine) checkDeregisteredHost(info *reconcileInfo) (result actionResult) {
+	// If the deregistered annotation is set we remove any host from the
+	// provisioner and take no further action
+	// Note this doesn't change the current state, only the OperationalStatus
+	annotations := hsm.Host.GetAnnotations()
+	if annotations != nil {
+		if _, ok := annotations[metal3v1alpha1.DeregisteredAnnotation]; ok {
+			if info.host.OperationalStatus() != metal3v1alpha1.OperationalStatusDeregistered {
+				_, err := hsm.Provisioner.Delete()
```

Comment:
Note: adding a 
```
if actionRes := hsm.ensureProvisioningCapacity(info); actionRes != nil {
			return actionRes
}
```
just before the `Provisioner.Delete()` will be enough to check for a free slot before addressing the removal.

Reply:
I don't see any reason to check for a free slot. We don't do that for any other calls that only touch ironic.

Reply:
Right, but in #807 the `nodes.Deleting` state is going to be included, so a node that is currently being deregistered will be considered occupying a slot.

Reply:
(note that I do not consider this a blocking point for the current PR, but I was noticing that the thanks to the latest @hardys refactoring it would have been simpler to add such check if required)

Reply:
Ah, I see that you are another victim of terrible naming in ironic.
A reasonable person might think that nodes.Deleting is the state that the node goes into when you delete it. It isn't.

nodes.Deleting is what happens when you deprovision the node, i.e. a Nova delete in OpenStack. It's triggered by a HTTP PATCH setting targetState=deleting. In metal³ this happens in the Deprovision() call.
When you actually delete a node it just disappears from the database and that's the end of it. It's triggered by a HTTP DELETE. In metal³ this happens in the Delete() call.
These two completely unrelated and yet destructive actions both have the same name. It confused me for the longest time.

Reply:
This was discussed in an [earlier thread](https://github.com/metal3-io/baremetal-operator/pull/827#discussion_r599761953) and I'm not sure we achieved consensus re whether the Delete() is a sufficiently inexpensive operation to skip the ProvisioningCapacity check.

Personally I think it's probably fine - just deleting the node from the DB is a *way* less expensive operation than deprovisioning, and the main cause of failures is stuff like PXE timeouts not DB contention.

I think @s3rj1k and @andfasano were more in favor of adding such a check, sounds like @zaneb and I aren't so :man_shrugging: - perhaps we leave it out for now, and it can be easily added if we ever see evidence of DB contention etc?

Reply:
> Ah, I see that you are another victim of terrible naming in ironic.
> A reasonable person might think that nodes.Deleting is the state that the node goes into when you delete it. It isn't.
> 
> nodes.Deleting is what happens when you deprovision the node, i.e. a Nova delete in OpenStack. It's triggered by a HTTP PATCH setting targetState=deleting. In metal³ this happens in the Deprovision() call.
> When you actually delete a node it just disappears from the database and that's the end of it. It's triggered by a HTTP DELETE. In metal³ this happens in the Delete() call.
> These two completely unrelated and yet destructive actions both have the same name. It confused me for the longest time.

Doh! Never mind then :). @hardys no need to add the check then

Reply:
@hardys That depends on amount of hosts, connection and memory limits in mysql db, on big clusters this will be an issue.

Reply:
Ok sounds like consensus is to skip the check, we can revisit in future if needed